### PR TITLE
Some Cata fixes

### PR DESCRIPTION
**Pull request by Appled**
Units: Implemented power types for Cataclysm
Units: Rewrote power regeneration
Units: Rewrote mana regeneration calculations
Spells: Rewrote spell start and spell go packets
Players: Fix player corpses (typo in corpse.cpp)
Misc: Organize SpellHandlers library
Misc: Fix crash with undefined spell aura effect
Misc: Clean C style casts

Closes https://github.com/AscEmu/AscEmu/issues/705
Closes https://github.com/AscEmu/AscEmu/issues/586

### CODE REVIEW

File: src/world/Units/Unit.cpp [hunk 1]
```diff
@@ -76,167 +77,409 @@ void Unit::setClass(uint8_t class_) { write(unitData()->field_bytes_0.s.unit_cla
...
+    if (type == POWER_TYPE_HEALTH)
+        return getHealth();
+
+    // Since cata power fields work differently
+    // Get matching power index by power type
+    const auto powerIndex = getPowerIndexFromDBC(type);
+    switch (powerIndex)
     {
-        case POWER_TYPE_MANA:
+        case 1:
```
Comment (Zyres) [inline src/world/Units/Unit.cpp]: Attention, "magic numbers" are not the best solution if we could have some enum values. If there is no fitting enum for this usecase, we should create one.

File: src/world/Units/Unit.cpp [hunk 2]
```diff
@@ -76,167 +77,409 @@ void Unit::setClass(uint8_t class_) { write(unitData()->field_bytes_0.s.unit_cla
...
 void Unit::setGender(uint8_t gender) { write(unitData()->field_bytes_0.s.gender, gender); }
 
-uint8_t Unit::getPowerType() const { return unitData()->field_bytes_0.s.power_type; }
-void Unit::setPowerType(uint8_t powerType) { write(unitData()->field_bytes_0.s.power_type, powerType); }
+PowerType Unit::getPowerType() const { return static_cast<PowerType>(unitData()->field_bytes_0.s.power_type); }
+void Unit::setPowerType(uint8_t powerType)
+{
+    write(unitData()->field_bytes_0.s.power_type, powerType);
+
+#ifdef AE_TBC
```
Comment (Zyres) [inline src/world/Units/Unit.cpp]: Use basic VERSION_STRING check, AE_XYZ will be removed soon.

Comment (Appled) [reply src/world/Units/Unit.cpp]: I've updated them with VERSION_STRING checks

File: src/world/Units/Unit.cpp (hunk 1 above)
Comment (Appled) [reply src/world/Units/Unit.cpp]: I've created a new enum to UnitDefines

File: src/world/Spell/SpellAuras.Legacy.cpp [hunk 3]
```diff
@@ -369,8 +369,61 @@ pSpellAura SpellAuraHandler[TOTAL_SPELL_AURAS] =
...
+    &Aura::SpellAuraNULL,                                                   // 364
+    &Aura::SpellAuraNULL,                                                   // 365
+    &Aura::SpellAuraNULL,                                                   // 366
+    &Aura::SpellAuraNULL,                                                   // 367
+    &Aura::SpellAuraNULL,                                                   // 368
+    &Aura::SpellAuraNULL,                                                   // 369
+    &Aura::SpellAuraNULL                                                    // 370
 };
 
 const char* SpellAuraNames[TOTAL_SPELL_AURAS] =
```
[the remark this answers is not in the document]
Comment (Appled) [reply src/world/Spell/SpellAuras.Legacy.cpp]: Already done https://github.com/AscEmu/AscEmu/blob/master/src/world/Spell/Definitions/AuraEffects.h#L356

File: src/world/Spell/SpellEffects.Legacy.cpp [hunk 4]
```diff
@@ -235,6 +235,26 @@ pSpellEffect SpellEffectsHandler[TOTAL_SPELL_EFFECTS] =
...
+    &Spell::SpellEffectNULL,                    // 176
+    &Spell::SpellEffectNULL,                    // 177
+    &Spell::SpellEffectNULL,                    // 178
+    &Spell::SpellEffectNULL,                    // 179
+    &Spell::SpellEffectNULL,                    // 180
+    &Spell::SpellEffectNULL,                    // 181
+    &Spell::SpellEffectNULL                     // 182
 };
 
 const char* SpellEffectNames[TOTAL_SPELL_EFFECTS] =
```
[the remark this answers is not in the document]
Comment (Appled) [reply src/world/Spell/SpellEffects.Legacy.cpp]: Already done https://github.com/AscEmu/AscEmu/blob/master/src/world/Spell/Definitions/SpellEffects.h#L193

**Pull request merged**